tollway/utils.py

```diff
@@ -1,43 +1,70 @@
 import json
+import logging
+from collections import namedtuple
 from pathlib import Path
-from typing import Mapping, TypedDict, Union
+from typing import Callable, Mapping, TypedDict, Union
 
 from decouple import config
 from google.cloud import pubsub_v1
+from google.cloud.pubsub_v1.publisher.futures import Future
 from google.oauth2 import service_account
 
+Topic = namedtuple("Topic", ["publisher", "topic_path"])
+
 
 class EventsLog(TypedDict):
     late_events: dict[str, list[str]]
     past_events: list[dict[str, str]]
     all_events: list[Mapping[str, Union[str, bool]]]
 
 
-def get_topic(pubsub: bool) -> tuple:
+def get_topic(pubsub: bool) -> Topic:
     if pubsub:
         project = config("PROJECT_ID")
         topic = config("TOPIC_ID")
         credentials = service_account.Credentials.from_service_account_file(config("PUBSUB_SERVICE_ACCOUNT"))
         publisher = pubsub_v1.PublisherClient(credentials=credentials)
         topic_path = publisher.topic_path(project=project, topic=topic)
-        return publisher, topic_path
-    return None, None
+        return Topic(publisher=publisher, topic_path=topic_path)
+    return Topic(publisher=None, topic_path=None)
+
+
+def get_pubsub_logger(pubsub: bool) -> logging.Logger:
+    logger = logging.getLogger("pubsub")
+    if pubsub:
+        logging.basicConfig(
+            level=logging.INFO,
+            format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
+            filename="pubsub.log",
+            filemode="a",
+        )
+    else:
+        logger.handlers.clear()
+        logger.addHandler(logging.NullHandler())
+    return logger
 
 
 def encode_message(message: dict) -> bytes:
     return json.dumps(message).encode("utf-8")
 
 
+def future_callback(logger: logging.Logger) -> Callable[[Future], None]:
+    def callback(future: Future) -> None:
+        try:
+            message_id = future.result()
+            logger.info(f"Message published with ID: {message_id}")
+        except Exception as e:
+            logger.error(f"An exception occurred: {e}")
+
+    return callback
+
+
 def write_to_file(filename: str, events_log: list[Mapping[str, Union[str, bool]]]):
     path = Path().absolute()
     with open(path / filename, mode="a") as file:
         json.dump(obj=events_log, fp=file, indent=1)
 
 
-def get_envs(env: str) -> int:
-    defaults = {"duplicate": 50, "all": 250, "seconds": 10, "minutes": 20, "hours": 30, "days": 100}
-
-    env_variable = config(env.upper(), default=defaults[env])
-    if env_variable == "":
-        return defaults[env]
-    return env_variable
+def get_config_with_default(key: str, default_value: str) -> int:
+    value = config(key, default=default_value)
+    return int(default_value) if value.strip() == "" else int(value)
```

Comment:
seems you can do one int() cast here

Reply:
yeah, I was testing a few different scenarios:

1. No .env file
2. .env file with few of variables defined
3. .env file with all variables defined

I kept hitting value int() doesn't have .strip() attribute. I can't remember exactly.

Reply:
so I was hitting value int() type doesn't have .strip() but I think that was for something else I was trying out. I think I originally had `default_value` parameter receiving an argument of type `int`. 

Anyways, if I don't place an `int` in the else block then what is returned is a string and that won't work against the len() comparison in the `events` module. Not sure if you can see, but here's a screenshot showing DUPLICATE_RATE defined in .env as 10, remove `int` from else block and "10" is returned. 

<img width="1929" alt="Screenshot 2024-02-14 at 5 14 45 PM" src="https://github.com/BenGriffith/tollway-traffic/assets/6089155/28bfa401-d2f9-4336-bfda-df12117c0b00">

Reply:
hm interesting, I remember now that config has a cast kwarg, e.g. `DEBUG = config('DEBUG', default=False, cast=bool)`, would an `cast=int` help here?